File: Registration-System-For-Students-in-Java-main/Registration System For Students in Java/course_section.py
import random
from course import Course
from mandatory_course import MandatoryCourse

from schedule import Schedule

class CourseSection(object):
    
    def __init__(self, course: Course):
        self.course = course
        self.course_program = []
        self.students = []
        
        
        
    def set_course_program(self):
        section_hours = self.course.total_hours()
        
        for i in range(section_hours):
            rand_hour = random.randint(0, Schedule.HOURS - 1)
            rand_day = random.randint(0, Schedule.DAYS - 1)
            
            if not self.course_program[rand_hour][rand_day] and not self.collides_with_same_semester(rand_hour, rand_day):
                self.course_program[rand_hour][rand_day] = True
            else:
                i -= 1        
                
                
    def collides_with_same_semester(self, rand_hour, rand_day):
        if not isinstance(self.course, MandatoryCourse):
            return False
        
        mandatory_courses = RegistrationSystem.mandatory_courses
        return any(self.course.semester_num == c.semester_number and  c.course_section.course_program()[rand_hour][rand_day] 
                   for c in mandatory_courses)
        
    
    def is_full(self):
        return len(self.students) >= self.course.quota

File: Registration-System-For-Students-in-Java-main/Registration System For Students in Java/iteration 2/src/CourseSection.java
import java.util.ArrayList;

public class CourseSection {

    private Course course;
    private RegistrationSystem registrationSystem;
    private boolean full;
    private int sectionHour;
    private ArrayList<Student> students;
    private boolean[][] courseProgram;

    public CourseSection(Course course) {
        this.course = course;
        setSectionHour();
        this.registrationSystem = RegistrationSystem.getInstance();
        students = new ArrayList<>();
        setFull();
        courseProgram = new boolean[Schedule.HOURS][Schedule.DAYS];
        setCourseProgram();
        //registrationSystem.getCourseSections().add(this);
    }

    /**Sets the courseProgram by adding all the lecture
     * hours to a random day and hour*/
    public void setCourseProgram() {
        for (int i = 0; i < sectionHour; i++) {
            int randomHour = (int)(Math.random() * Schedule.HOURS);
            int randomDay = (int)(Math.random() * Schedule.DAYS);

            //If course program is empty for that hour and same semester course has no lectures in that hour
            if (!courseProgram[randomHour][randomDay] && !CollidesWithSameSemester(randomHour, randomDay)) {
                courseProgram[randomHour][randomDay] = true;
            }else {
                i--;
            }
        }
    }

    private boolean CollidesWithSameSemester(int randomHour, int randomDay) {
        if (!(course instanceof MandatoryCourse)) {
            return false;
        }

        for (MandatoryCourse c: registrationSystem.getMandatoryCourses()) {
            if (((MandatoryCourse) course).getSemesterNumber() == c.getSemesterNumber() &&
                    c.getCourseSection().getCourseProgram()[randomHour][randomDay]) {
                return true;
            }
        }

        return false;
    }

    public boolean addStudent(Student student) {
        if (!full) {
            students.add(student);
            student.addToCurrentCourses(this);
            setFull();
            return true;
        }else {
            student.getExecutionTrace().append("\nThe system didn't allow " + course.toString() + " because " +
                    "course section is full. ("  +  students.size() + ")");
            course.getNonRegisteredQuota().add(student);
            return false;
        }
    }

    public int getQuota() {
        return course.getQuota();
    }

    public boolean isFull() {
        return full;
    }

    public Course getCourse() {
        return course;
    }

    public String getCourseSectionCode() {
        return getCourse().getCourseCode();
    }

    public void setFull(boolean full) {
        this.full = full;
    }

    public void setFull() {
        full = students.size() == getQuota();
    }



    public int getSectionHour() {
        return sectionHour;
    }

    public void setSectionHour() {
        sectionHour = course.getSectionHours();
    }

    public ArrayList<Student> getStudents() {
        return students;
    }

    public void setCourse(Course course) {
        this.course = course;
    }


    public boolean[][] getCourseProgram() {
        return courseProgram;
    }

}
